src/lib/kafka/KafkaConnectionManager.cpp

```diff
@@ -0,0 +1,404 @@
+* published by the Free Software Foundation, either version 3 of the
+* License, or (at your option) any later version.
+*
+* Orion Context Broker is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
+* General Public License for more details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+*
+* For those usages not covered by this license please contact with
+* iot_support at tid dot es
+*
+* Author: Fermin Galan (with contributions from Burak Karaboga in old mqtt.cpp file)
+*/
+
+#include "kafka/KafkaConnectionManager.h"
+
+#include <iostream>
+#include <mosquitto.h>
+#include <string>
+#include <vector>
+#include <map>
+
+#include "logMsg/logMsg.h"
+#include "common/globals.h"
+#include "alarmMgr/alarmMgr.h"
+
+#include <librdkafka/rdkafka.h>
+#include <stdio.h>
+#include <signal.h>
+#include <string.h>
+
+
+/* ****************************************************************************
+*
+* KAFKA_DEFAULT_KEEPALIVE -
+*/
+#define KAFKA_DEFAULT_KEEPALIVE 60
+
+
+
+/* ****************************************************************************
+*
+* getEndpoint -
+*/
+inline std::string getEndpoint(const std::string& host, int port)
+{
+  char  portV[STRING_SIZE_FOR_INT];
+  snprintf(portV, sizeof(portV), "%d", port);
+  return host + ":" + portV;
+}
+
+
+
+/* ****************************************************************************
+*
+* KafkaConnectionManager::KafkaConnectionManager -
+*/
+KafkaConnectionManager::KafkaConnectionManager(void)
+{
+}
+
+
+
+/* ****************************************************************************
+*
+* KafkaConnectionManager::init -
+*/
+void KafkaConnectionManager::init(long _timeout)
+{
+  LM_T(LmtMqttNotif, ("Initializing KAFKA library"));
+  mosquitto_lib_init();
+
+  if (_timeout != -1)
+  {
+    timeout = _timeout;
+  }
+  else
+  {
+    timeout = DEFAULT_TIMEOUT;
+  }
+
+  semInit();
+}
+
+
+
+/* ****************************************************************************
+*
+* KafkaConnectionManager::teardown -
+*/
+void KafkaConnectionManager::teardown(void)
+{
+  LM_T(LmtMqttNotif, ("Teardown KAFKA connections"));
+
+  for (std::map<std::string, KafkaConnection*>::iterator iter = connections.begin(); iter != connections.end(); ++iter)
+  {
+    std::string endpoint = iter->first;
+    KafkaConnection* cP   = iter->second;
+
+    disconnect(cP->producer, endpoint);
+    cP->producer = NULL;
+
+    delete cP;
+  }
+  connections.clear();
+
+  // 5. Limpieza global de librdkafka (esperar hasta 3 segundos)
+  int wait_time_ms = 3000;
+  while (rd_kafka_wait_destroyed(wait_time_ms) == -1)
+  {
+    LM_I(("Esperando finalización de threads internos de librdkafka..."));
+    wait_time_ms -= 100;
+    if (wait_time_ms <= 0) break;
+  }
```

Comment:
Avoid text/comments in Spanish

Avoid direct usage of 3000 ("magic number"). Use a constant declared in a #define